pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotJoinToDynamicBroadcastRule.java

```diff
@@ -0,0 +1,171 @@
+ *   @see <a href="https://github.com/apache/pinot/pull/10565/>
+ * TODO #2: Only convert to dynamic broadcast from right-to-left, allow option to specify dynamic broadcast direction.
+ * TODO #3: Only convert to dynamic broadcast if left is leaf stage, allow the option for intermediate stage.
+ * TODO #4: currently adding a pass-through after join, support leaf-stage to chain arbitrary operator(s) next.
+ */
+public class PinotJoinToDynamicBroadcastRule extends RelOptRule {
+  public static final PinotJoinToDynamicBroadcastRule INSTANCE =
+      new PinotJoinToDynamicBroadcastRule(PinotRuleUtils.PINOT_REL_FACTORY);
+  private static final String DYNAMIC_BROADCAST_HINT_OPTION_VALUE = "dynamic_broadcast";
+
+  public PinotJoinToDynamicBroadcastRule(RelBuilderFactory factory) {
+    super(operand(LogicalJoin.class, any()), factory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    if (call.rels.length < 1 || !(call.rel(0) instanceof Join)) {
+      return false;
+    }
+    Join join = call.rel(0);
+    String joinStrategyString = PinotHintStrategyTable.getHintOption(join.getHints(),
+        PinotHintOptions.JOIN_HINT_OPTIONS, PinotHintOptions.JoinHintOptions.JOIN_STRATEGY);
+    List<String> joinStrategies = joinStrategyString != null ? StringUtils.split(joinStrategyString, ",")
+        : Collections.emptyList();
+    if (!joinStrategies.contains(DYNAMIC_BROADCAST_HINT_OPTION_VALUE)) {
+      return false;
+    }
+    JoinInfo joinInfo = join.analyzeCondition();
+    RelNode left = join.getLeft() instanceof HepRelVertex ? ((HepRelVertex) join.getLeft()).getCurrentRel()
+        : join.getLeft();
+    RelNode right = join.getRight() instanceof HepRelVertex ? ((HepRelVertex) join.getRight()).getCurrentRel()
+        : join.getRight();
+    return left instanceof LogicalExchange && right instanceof LogicalExchange
+        && PinotRuleUtils.noExchangeInSubtree(left.getInput(0))
+        && (join.getJoinType() == JoinRelType.SEMI && joinInfo.nonEquiConditions.isEmpty());
+  }
+
+  @Override
+  public void onMatch(RelOptRuleCall call) {
+    Join join = call.rel(0);
+    LogicalExchange left = (LogicalExchange) (join.getLeft() instanceof HepRelVertex
+        ? ((HepRelVertex) join.getLeft()).getCurrentRel() : join.getLeft());
+    LogicalExchange right = (LogicalExchange) (join.getRight() instanceof HepRelVertex
+        ? ((HepRelVertex) join.getRight()).getCurrentRel() : join.getRight());
+
+    // TODO: both optimized hash empty list should be singleton exchange. however it is broken. fix later.
+    boolean isColocatedJoin = PinotHintStrategyTable.containsHintOption(join.getHints(),
+        PinotHintOptions.JOIN_HINT_OPTIONS, PinotHintOptions.JoinHintOptions.IS_COLOCATED_BY_JOIN_KEYS);
+    LogicalExchange dynamicBroadcastExchange = isColocatedJoin
+        ? LogicalExchange.create(right.getInput(), RelDistributions.hash(Collections.emptyList()))
```

Comment:
TODO: we have to move this to hash with empty List (deterministically to 1-server) instead of singleton. we need to clean up singleton before this can work. CC @xiangfu0

Reply:
singleton fix pending: https://github.com/apache/pinot/pull/10797